src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs

```diff
@@ -461,22 +474,41 @@ protected virtual void ReportUnassignedOutParameter(ParameterSymbol parameter, S
             MethodSymbol member,
             BoundNode node,
             DiagnosticBag diagnostics,
-            bool requireOutParamsAssigned = true)
+            out ImmutableArray<FieldSymbol> implicitlyInitializedFieldsOpt,
+            bool requireOutParamsAssigned)
         {
             Debug.Assert(diagnostics != null);
 
             // Run the strongest version of analysis
-            DiagnosticBag strictDiagnostics = analyze(strictAnalysis: true);
+            (DiagnosticBag strictDiagnostics, implicitlyInitializedFieldsOpt) = analyze(strictAnalysis: true);
             if (strictDiagnostics.IsEmptyWithoutResolution)
             {
                 // If it reports nothing, there is nothing to report and we are done.
                 strictDiagnostics.Free();
+                Debug.Assert(implicitlyInitializedFieldsOpt.IsDefault);
                 return;
             }
 
             // Also run the compat (weaker) version of analysis to see if we get the same diagnostics.
             // If any are missing, the extra ones from the strong analysis will be downgraded to a warning.
-            DiagnosticBag compatDiagnostics = analyze(strictAnalysis: false);
+            (DiagnosticBag compatDiagnostics, var compatImplicitlyInitializedFieldsOpt) = analyze(strictAnalysis: false);
+
+            // note that "compat" may be missing some implicit initializations, because it considers certain fields to always be fully assigned.
+            Debug.Assert(compatImplicitlyInitializedFieldsOpt.IsDefault
+                || (implicitlyInitializedFieldsOpt is var strictFields && compatImplicitlyInitializedFieldsOpt.All(field => strictFields.Contains(field))));
+
+            if (MessageID.IDS_FeatureImplicitInitializationInStructConstructors.GetFeatureAvailabilityDiagnosticInfo(compilation) is { } diagnosticInfo)
+            {
+                // We only give the LangVersion error if "compat" initializations are present.
+                if (!compatImplicitlyInitializedFieldsOpt.IsDefault)
```

Comment:
I didn't understand why we'll auto-default the fields listed in `implicitlyInitializedFieldsOpt`, but we'll only produce a LangVer diagnostic based on the fields listed in `compatImplicitlyInitializedFieldsOpt`. I would have expected a LangVer check any time we had to do an auto-default for a field.

Reply:
See WarningVersionTests.cs for places where these behaviors are exercised.

Idea here is that we preserve the existing behaviors pre-C# 11, by never inserting the redundant initializations, and by only giving LangVersion diagnostics when an initialization *would* be used with the "compat" rules.

However, moving from C# 10 to C# 11 *can* result in a silent change in codegen here. e.g.

```cs
using System.Collections.Immutable;

public struct S
{
    public ImmutableArray<int> _arr;

    public S(bool unused) { }
}
```

In the above sample, we:
1. give no diagnostics at all here pre-.NET 5. We just silently let you use uninitialized memory at the call site of the constructor here.
2. in .NET 5, we give a "warning wave" warning that _arr is not initialized, but compile and emit exactly the same way.
3. in .NET 7 (C# 11), we give a disabled-by-default warning, and implicitly insert `_arr = default;` into the constructor.